File: modules/relisten-audio-player/ios/RelistenAudioPlayerModule.swift
import ExpoModulesCore
import Foundation

private let moduleLifecycleLog = RelistenPlaybackLogger(layer: .backend, category: .lifecycle)

struct RelistenStreamable: Record {
    @Field
    var url: URL?

    @Field
    var identifier: String?

    @Field
    var cacheKey: String?

    @Field
    var title: String?

    @Field
    var albumTitle: String?

    @Field
    var albumArt: String?

    @Field
    var artist: String?

    @Field
    var downloadDestination: URL?
}

var DEBUG_state = ""

public class RelistenAudioPlayerModule: Module {
    var player: PlaybackBackend?
    private let moduleInstanceID = UUID().uuidString

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    public func definition() -> ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('RelistenAudioPlayer')` in JavaScript.
        Name("RelistenAudioPlayer")

        OnCreate {
            moduleLifecycleLog.info(
                "created",
                "module instance",
                playbackLogField("id", moduleInstanceID)
            )
            player = makePlaybackBackend()
            player?.delegate = self
        }
        
        OnDestroy {
            moduleLifecycleLog.info(
                "destroyed",
                "module instance",
                playbackLogField("id", moduleInstanceID)
            )
            player?.teardown()
            player = nil
        }

        Events(
            "onError",
            "onPlaybackStateChanged",
            "onPlaybackProgressChanged",
            "onDownloadProgressChanged",
            "onTrackChanged",
            "onRemoteControl",
            "onTrackStreamingCacheComplete"
        )

        Function("currentDuration") {
            return player?.currentDurationSnapshot
        }

        Function("currentState") {
            return player?.currentStateSnapshot ?? .Stopped
        }

        Function("currentStateStr") {
            return player?.currentStateString ?? String(describing: PlaybackState.Stopped)
        }

        Function("elapsed") {
            return player?.elapsedSnapshot
        }

        Function("volume") {
            return player?.volume ?? 0.0
        }

        Function("setVolume") { (newVolume: Float) in
            return player?.volume = newVolume
        }

        Function("prepareAudioSession") {
            player?.enqueuePrepareAudioSession()
        }

        AsyncFunction("playbackProgress") { (promise: Promise) in
            guard let player = self.player else {
                promise.resolve(["playbackProgress": nil] as [String: Any?])
                return
            }

            player.requestPlaybackProgress { snapshot in
                promise.resolve([
                    "playbackProgress": [
                        "elapsed": snapshot.elapsed,
                        "duration": snapshot.duration
                    ] as [String: TimeInterval?],
                    "activeTrackDownloadProgress": [
                        "forActiveTrack": true,
                        "downloadedBytes": snapshot.activeTrackDownloadedBytes as Any,
                        "totalBytes": snapshot.activeTrackTotalBytes as Any
                    ] as [String: Any]
                ] as [String: Any])
            }
        }

        AsyncFunction("play") { (streamable: RelistenStreamable, startingAtMs: Int64?, promise: Promise) in
            guard let url = streamable.url,
                  let identifier = streamable.identifier,
                  let cacheKey = streamable.cacheKey,
                  let title = streamable.title,
                  let albumArt = streamable.albumArt,
                  let albumTitle = streamable.albumTitle,
                  let artist = streamable.artist else {
                promise.resolve()
                return
            }

            player?.enqueuePlay(
                RelistenGaplessStreamable(
                    url: url,
                    identifier: identifier,
                    cacheKey: cacheKey,
                    title: title,
                    artist: artist,
                    albumTitle: albumTitle,
                    albumArt: albumArt,
                    downloadDestination: streamable.downloadDestination
                ),
                startingAtMs: startingAtMs
            ) {
                promise.resolve()
            }
        }

        Function("setNextStream") { (streamable: RelistenStreamable?) in
            if streamable == nil {
                player?.enqueueSetNextStream(nil)
                return
            }

            guard let streamable = streamable,
                  let url = streamable.url,
                  let identifier = streamable.identifier,
                  let cacheKey = streamable.cacheKey,
                  let title = streamable.title,
                  let albumTitle = streamable.albumTitle,
                  let albumArt = streamable.albumArt,
                  let artist = streamable.artist else {
                return
            }
            
            player?.enqueueSetNextStream(
                RelistenGaplessStreamable(
                    url: url,
                    identifier: identifier,
                    cacheKey: cacheKey,
                    title: title,
                    artist: artist,
                    albumTitle: albumTitle,
                    albumArt: albumArt,
                    downloadDestination: streamable.downloadDestination
                )
            )
        }

        Function("setRepeatMode") { (repeatMode: Int) in
            player?.enqueueSetRepeatMode(repeatMode)
        }

        Function("setShuffleMode") { (shuffleMode: Int) in
            player?.enqueueSetShuffleMode(shuffleMode)
        }

        AsyncFunction("resume") { (promise: Promise) in
            player?.enqueueResume {
                promise.resolve()
            }
        }

        AsyncFunction("pause") { (promise: Promise) in
            player?.enqueuePause {
                promise.resolve()
            }
        }

        AsyncFunction("stop") { (promise: Promise) in
            player?.enqueueStop {
                promise.resolve()
            }
        }

        AsyncFunction("next") { (promise: Promise) in
            player?.enqueueNext {
                promise.resolve()
            }
        }

        AsyncFunction("seekTo") { (pct: Double, promise: Promise) in
            player?.enqueueSeekTo(percent: pct) {
                promise.resolve()
            }
        }
        
        AsyncFunction("seekToTime") { (timeMs: Int64, promise: Promise) in
            player?.enqueueSeekToTime(timeMs) {
                promise.resolve()
            }
        }
    }
}

extension RelistenAudioPlayerModule: PlaybackBackendDelegate {
    public func sendAndLogEvent(_ eventName: String, _ body: [String: Any?] = [:]) {
        NSLog("[relisten-audio-player] sendEvent: eventName=\(eventName) body=\(body)")
        self.sendEvent(eventName, body)
    }
    
    public func streamingCacheCompleted(forStreamable streamable: RelistenGaplessStreamable, bytesWritten: Int) {
        self.sendAndLogEvent("onTrackStreamingCacheComplete", [
            "identifier": streamable.identifier,
            "totalBytes": bytesWritten
        ])
    }
    
    public func errorStartingStream(error: NSError, forStreamable: RelistenGaplessStreamable) {
        self.sendAndLogEvent("onError", [
            "error": error.code,
            "errorDescription": error.localizedDescription,
            "identifier": forStreamable.identifier
        ])
    }

    public func playbackStateChanged(newPlaybackState playbackState: PlaybackState) {
        self.sendAndLogEvent("onPlaybackStateChanged", [
            "newPlaybackState": String(describing: playbackState)
        ])
    }

    public func playbackProgressChanged(elapsed: TimeInterval?, duration: TimeInterval?) {
        self.sendEvent("onPlaybackProgressChanged", [
            "elapsed": elapsed,
            "duration": duration
        ])
    }

    public func downloadProgressChanged(forActiveTrack: Bool, downloadedBytes: UInt64, totalBytes: UInt64) {
        self.sendEvent("onDownloadProgressChanged", [
            "forActiveTrack": forActiveTrack,
            "downloadedBytes": downloadedBytes,
            "totalBytes": totalBytes
        ])
    }

    public func trackChanged(previousStreamable: RelistenGaplessStreamable?, currentStreamable: RelistenGaplessStreamable?) {
        self.sendAndLogEvent("onTrackChanged", [
            "previousIdentifier": previousStreamable?.identifier,
            "currentIdentifier": currentStreamable?.identifier
        ])
    }

    public func remoteControl(method: String) {
        self.sendAndLogEvent("onRemoteControl", [
            "method": method
        ])
    }

    public func audioSessionWasSetup() {

    }
}

File: modules/relisten-audio-player/android/src/main/java/net/relisten/android/audio_player/RelistenAudioPlayerModule.kt
package net.relisten.android.audio_player

import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.launch
import net.relisten.android.audio_player.gapless.RelistenGaplessAudioPlayer
import net.relisten.android.audio_player.gapless.RelistenGaplessAudioPlayerDelegate
import net.relisten.android.audio_player.gapless.RelistenGaplessStreamable
import net.relisten.android.audio_player.gapless.RelistenPlaybackState
import net.relisten.android.audio_player.gapless.internal.RelistenPlaybackException
import java.net.URL

class RelistenStreamable : Record {
    @Field
    var url: URL? = null

    @Field
    var identifier: String? = null

    @Field
    var title: String? = null

    @Field
    var albumTitle: String? = null

    @Field
    var albumArt: String? = null

    @Field
    var artist: String? = null

    @Field
    var downloadDestination: URL? = null

    fun toGaplessStreamable(): RelistenGaplessStreamable? {
        if (url != null && identifier != null && title != null && albumTitle != null && albumArt != null && artist != null) {
            return RelistenGaplessStreamable(url!!, identifier!!, title!!, artist!!, albumTitle!!, albumArt!!, downloadDestination)
        }

        return null
    }
}

class RelistenAudioPlayerModule : Module(), RelistenGaplessAudioPlayerDelegate {
    var player: RelistenGaplessAudioPlayer? = null

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module' class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from requireNativeModule('RelistenAudioPlayer') -> JavaScript.
        Name("RelistenAudioPlayer")

        OnCreate {
            player = RelistenGaplessAudioPlayer(appContext)
            player?.delegate = this@RelistenAudioPlayerModule
        }

        OnDestroy {
            val thisPlayer = player

            if (thisPlayer != null) {
                thisPlayer.stop()
                thisPlayer.teardown()
            }

            player = null
        }

        Events("onError", "onPlaybackStateChanged", "onPlaybackProgressChanged", "onDownloadProgressChanged", "onTrackChanged")

        Function("currentDuration") {
            return@Function player?.currentDuration
        }

        Function("currentState") {
            return@Function player?.currentState ?: RelistenPlaybackState.Stopped
        }

        Function("currentStateStr") {
            return@Function (player?.currentState ?: RelistenPlaybackState.Stopped).toString()
        }

        Function("elapsed") {
            return@Function player?.elapsed
        }

        Function("volume") {
            return@Function player?.volume ?: 0.0
        }

        Function("setVolume") { newVolume: Float ->
            player?.volume = newVolume
        }

        Function("prepareAudioSession") {
            player?.prepareAudioSession()
        }

        AsyncFunction("playbackProgress") { promise: Promise ->
            player?.scope?.launch {
                val player = player

                if (player == null) {
                    promise.resolve(hashMapOf("playbackProgress" to null))
                    return@launch
                }

                promise.resolve(
                    hashMapOf(
                        "playbackProgress" to hashMapOf(
                            "elapsed" to player.elapsed,
                            "duration" to player.currentDuration,
                        ),
                        "activeTrackDownloadProgress" to hashMapOf(
                            "forActiveTrack" to true,
                            "downloadedBytes" to player.activeTrackDownloadedBytes as Any,
                            "totalBytes" to player.activeTrackTotalBytes as Any,
                        )
                    )
                )
            }
        }

        AsyncFunction("play") { streamable: RelistenStreamable, startingAtPct: Double?, promise: Promise ->
            player?.scope?.launch {
                val gaplessStreamable = streamable.toGaplessStreamable()

                if (gaplessStreamable != null) {
                    Log.e("relisten-audio", "player?.play")
                    player?.play(gaplessStreamable)
                    Log.e("relisten-audio", "after player?.play")
                }

                Log.e("relisten-audio", "promise.resolve")
                promise.resolve(null)
                Log.e("relisten-audio", "after promise.resolve")
            }
        }

        Function("setNextStream") { streamable: RelistenStreamable? ->
            if (streamable == null) {
                player?.setNextStream(null)
                return@Function
            }

            val gaplessStreamable = streamable.toGaplessStreamable()

            if (gaplessStreamable != null) {
                player?.setNextStream(gaplessStreamable)
            }

            return@Function
        }

        AsyncFunction("resume") { promise: Promise ->
            player?.scope?.launch {
                player?.resume()
                promise.resolve(null)
            }
        }

        AsyncFunction("pause") { promise: Promise ->
            player?.scope?.launch {
                player?.pause()
                promise.resolve(null)
            }
        }

        AsyncFunction("stop") { promise: Promise ->
            player?.scope?.launch {
                player?.stop()
                promise.resolve(null)
            }
        }

        AsyncFunction("next") { promise: Promise ->
            player?.scope?.launch {
                player?.next()
                promise.resolve(null)
            }
        }

        AsyncFunction("seekTo") { pct: Double, promise: Promise ->
            player?.scope?.launch {
                player?.seekTo(pct)
                promise.resolve(null)
            }
        }
    }

    override fun errorStartingStream(
            player: RelistenGaplessAudioPlayer,
            error: RelistenPlaybackException,
            forStreamable: RelistenGaplessStreamable
    ) {
        if (this.player == null) {
            return
        }

        sendEvent(
            "onError", hashMapOf(
                "error" to error.code,
                "errorMessage" to error.message,
                "errorDescription" to error.description,
                "identifier" to forStreamable.identifier,
            )
        )
    }

    override fun playbackStateChanged(
        player: RelistenGaplessAudioPlayer,
        newPlaybackState: RelistenPlaybackState
    ) {
        if (this.player == null) {
            return
        }

        sendEvent(
            "onPlaybackStateChanged", hashMapOf(
                "newPlaybackState" to newPlaybackState.toString(),
            )
        )
    }

    override fun playbackProgressChanged(
        player: RelistenGaplessAudioPlayer,
        elapsed: Double?,
        duration: Double?
    ) {
        if (this.player == null) {
            return
        }

        sendEvent(
            "onPlaybackProgressChanged", hashMapOf(
                "elapsed" to elapsed,
                "duration" to duration,
            )
        )
    }

    override fun downloadProgressChanged(
        player: RelistenGaplessAudioPlayer,
        forActiveTrack: Boolean,
        downloadedBytes: Long,
        totalBytes: Long
    ) {
        if (this.player == null) {
            return
        }

        sendEvent(
            "onDownloadProgressChanged", hashMapOf(
                "forActiveTrack" to forActiveTrack,
                "downloadedBytes" to downloadedBytes,
                "totalBytes" to totalBytes,
            )
        )
    }

    override fun trackChanged(
        player: RelistenGaplessAudioPlayer,
        previousStreamable: RelistenGaplessStreamable?,
        currentStreamable: RelistenGaplessStreamable?
    ) {
        if (this.player == null) {
            return
        }

        sendEvent(
            "onTrackChanged", hashMapOf(
                "previousIdentifier" to previousStreamable?.identifier,
                "currentIdentifier" to currentStreamable?.identifier,
            )
        )
    }

    override fun audioSessionWasSetup(player: RelistenGaplessAudioPlayer) {

    }
}
